### PR TITLE
[C23] Complete support for WG14 N2508

### PR DESCRIPTION
**Pull request by AaronBallman**
In Clang 16, we implemented the ability to add a label at the end of a compound statement. These changes complete the implementation by allowing a label to be followed by a declaration in C.

Note, this seems to have fixed an issue with some OpenMP stand-alone directives not being properly diagnosed as per:
https://www.openmp.org/spec-html/5.1/openmpsu19.html#x34-330002.1.3
(The same requirement exists in OpenMP 5.2 as well.)

### CODE REVIEW

File: clang/include/clang/Basic/DiagnosticParseKinds.td [hunk 1]
```diff
@@ -299,6 +299,12 @@ def note_missing_selector_name : Note<
 def note_force_empty_selector_name : Note<
   "or insert whitespace before ':' to use %0 as parameter name "
   "and have an empty entry in the selector">;
+def ext_c_label_followed_by_declaration : ExtWarn<
+  "label followed by a declaration is a C23 extension">,
+  InGroup<C23>;
+def warn_c23_compat_label_followed_by_declaration : Warning<
+  "label followed by a declaration is incompatible with C standards before "
+  "C23">, InGroup<CPre23Compat>, DefaultIgnore;
```
Comment (Fznamznon) [inline clang/include/clang/Basic/DiagnosticParseKinds.td:307]: I can't find tests that would check that these messages are emitted. Are they coming?

Comment (AaronBallman) [reply clang/include/clang/Basic/DiagnosticParseKinds.td:307]: Good catch, those slipped through the cracks. I'll add the tests shortly.

Comment (AaronBallman) [reply clang/include/clang/Basic/DiagnosticParseKinds.td:307]: This should be handled now.

File: clang/test/OpenMP/error_message.cpp [hunk 2]
```diff
@@ -32,8 +32,8 @@ T tmain(T argc) {
   }
   switch (argc) {
 #pragma omp error // expected-error {{ERROR}}
-  case 1:
-#pragma omp error // expected-error {{ERROR}}
+  case 1: // FIXME: error without 'at execution' is not a stand-alone directive and so this should be accepted.
```
Comment (mikerice1969) [inline clang/test/OpenMP/error_message.cpp:35]: I would remove this FIXME. I think we want 'error' to be diagnosed. The original meaning of stand-alone was directives that did not has associated user statements. This was before utility/informational directives existed. I think the intention is we only allow executable statements with associated user statements here, not declarative, utility, etc.

Comment (AaronBallman) [reply clang/test/OpenMP/error_message.cpp]: SGTM, thank you! This should be done now.

File: clang/lib/Parse/ParseStmt.cpp [hunk 3]
```diff
@@ -712,9 +722,7 @@ StmtResult Parser::ParseLabeledStatement(ParsedAttributes &Attrs,
   assert(Tok.is(tok::identifier) && Tok.getIdentifierInfo() &&
          "Not an identifier!");
 
-  // The substatement is always a 'statement', not a 'declaration', but is
-  // otherwise in the same context as the labeled-statement.
-  StmtCtx &= ~ParsedStmtContext::AllowDeclarationsInC;
+  StmtCtx &= ~ParsedStmtContext::AllowStandaloneOpenMPDirectives;
```
[the remark this answers is not in the document]
Comment (AaronBallman) [reply clang/lib/Parse/ParseStmt.cpp:725]: Done

**Review by Fznamznon — COMMENTED** [review]
LGTM, but someone else must approve.

**Comment by AaronBallman** [comment]
Ping

**Review by mikerice1969 — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**